controller/lib/hal/hal_stm32.h

```diff
@@ -0,0 +1,385 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This file implements the HAL (Hardware Abstraction Layer) for the
+STM32L452 processor used on the controller.  Details of the processor's
+peripherals can be found in the reference manual for that processor:
+   https://www.st.com/resource/en/reference_manual/dm00151940-stm32l41xxx42xxx43xxx44xxx45xxx46xxx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf
+
+Details specific to the ARM processor used in this chip can be found in
+the programmer's manual for the processor available here:
+   https://www.st.com/resource/en/programming_manual/dm00046982-stm32-cortexm4-mcus-and-mpus-programming-manual-stmicroelectronics.pdf
+
+*/
+
+#ifndef DEF_INC_STM32
+#define DEF_INC_STM32
+
+#if !defined(BARE_STM32)
+#error                                                                         \
+    "the header stm32.h is part of the HAL for the BARE_STM32 build.  It should only be used as part of that build"
+#endif
+
+#include <stdint.h>
+
+#define CPU_FREQ_MHZ 80
+#define CPU_FREQ (CPU_FREQ_MHZ * 1000000)
+
+// Some useful inline functions to enable/disable interrupts
+inline void IntDisable() { asm volatile("cpsid i" ::: "memory"); }
+
+inline void IntEnable() { asm volatile("cpsie i" ::: "memory"); }
+
+inline void IntRestore(bool p) {
+  if (p)
+    IntEnable();
+}
+
+inline bool IntSuspend() {
+  int ret;
+  asm volatile("mrs   %[output], primask\n\t"
+               "cpsid i"
+               : [output] "=r"(ret)::"memory");
+  return (ret == 0);
+}
+
+// Interrupt vectors that we currently use.
+// The values here are the offsets into the interrupt table.
+// These can be found in the NVIC chapter (chapter 12) of the
+// processor reference manual
+#define INT_VEC_UART3 0x0DC
+#define INT_VEC_TIMER6 0x118
+
+// Represents a 32-bit register
+typedef volatile uint32_t REG;
+
+// 16-bit short register
+typedef volatile uint16_t SREG;
+
+// 8-bit byte sized register
+typedef volatile uint8_t BREG;
+
+///////////////////////////////////////////////////////////////
+// The structures below represent the STM32 registers used
+// to configure various modules (like timers, serial ports, etc).
+// Detailed information on these modules and the registers
+// used to configure them can be found in the reference
+// manual for this chip.
+///////////////////////////////////////////////////////////////
+
+// Reset & clock controller
+struct RCC_Regs {
```

Comment:
Should these structs be marked `__attribute__ ((packed))`? Doesn't matter for those that only consist of REG, but for those containing SREG and BREG it would matter.

We can also place some `static_assert(sizeof(YourStruct) == ..., "...")` at global scope below the structs for some extra safety.

Reply:
Presumably not; `packed` means that e.g. uint32_t will/can be misaligned, but surely we don't want that.

The static_asserts sound good in any case and will show whether there's a packing problem.

Reply:
They're all properly aligned by the design of the processor.